### ISSUE TITLE
Should set state by passing a callback

### ISSUE THREAD

**Issue by matchatype** [the-road-to-learn-react/the-road-to-learn-react#123]
The recommended way to update state is by passing a callback instead of an updater object directly because `setState` is actually asynchronous, for example:

```js
onSearchChange = e => {
  this.setState(() => ({ searchTerm: e.target.value }))
}
```
However, in this specific case, you have to pay attention to persist the synthetic event, since the object is being reused and [you cannot access the event asynchronously](https://reactjs.org/docs/events.html#event-pooling). You can persist the event by calling `event.persist()`:

```js
onSearchChange = e => {
  e.persist()
  this.setState(() => ({ searchTerm: e.target.value }))
}
```

**Comment by lcnogueira** [comment]
Actually, it's not necessary to use a callback function in this case. You should use a callback function when the **new state depends on the previous one**. For example, if the state were a counter, so you'd better use a callback function to update the state instead of an object :bulb:.

If you want to know more about it, take a look at this documentation section: https://reactjs.org/docs/faq-state.html#how-do-i-update-state-with-values-that-depend-on-the-current-state.

**Comment by matchatype** [comment]
I read...

> Pass a function instead of an object to setState to ensure the call always uses the most updated version of state (see below).

...and...

> Currently, `setState` is asynchronous inside event handlers.

```
// `e` is short for event.
onSearchChange = e => { ... }
```

**Comment by lcnogueira** [comment]
Exactly. You need to use a callback function if the new state depends on the most updated version of state. However, in this case, your new state doesn't depend on the most updated version of state. It only depends on what the user types in, because it always uses `e.target.value` to set the new state. 

In fact, `setState` is asynchronous inside event handlers. But there is no problem, because your new state is gonna be what the user types in. 

There is a reason why the documentation shows a counter to explain this: when you have a counter, the new value (new state) depends on the most updated value (most updated version of state). Ex.: if the counter value is 5, so the new state should be 6 (the most update version + 1).

**Comment by matchatype** [comment]
I'm sorry, I rushed at your comment previously. It's correct, in this simple example, we probably would not run into any issues by passing an object.
